File: client/src/test/scala/stasis/test/specs/unit/client/mocks/MockRecoveryMetadataCollector.scala
package stasis.test.specs.unit.client.mocks

import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

import scala.concurrent.Future

import stasis.client.collection.RecoveryMetadataCollector
import stasis.client.model.EntityMetadata
import stasis.client.model.TargetEntity
import stasis.test.specs.unit.client.mocks.MockRecoveryMetadataCollector.Statistic

class MockRecoveryMetadataCollector(metadata: Map[Path, EntityMetadata]) extends RecoveryMetadataCollector {
  private val stats: Map[Statistic, AtomicInteger] = Map(
    Statistic.FileCollectedWithExistingMetadata -> new AtomicInteger(0)
  )

  override def collect(
    file: Path,
    destination: TargetEntity.Destination,
    existingMetadata: EntityMetadata
  ): Future[TargetEntity] = {
    stats(Statistic.FileCollectedWithExistingMetadata).incrementAndGet()
    metadata.get(file) match {
      case Some(fileMetadata) =>
        Future.successful(
          TargetEntity(
            path = file,
            destination = destination,
            existingMetadata = existingMetadata,
            currentMetadata = Some(fileMetadata)
          )
        )

      case None =>
        Future.failed(new IllegalArgumentException(s"No metadata found for file [$file]"))
    }
  }

  def statistics: Map[Statistic, Int] = stats.view.mapValues(_.get()).toMap
}

object MockRecoveryMetadataCollector {
  sealed trait Statistic
  object Statistic {
    case object FileCollectedWithExistingMetadata extends Statistic
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockRecoveryMetadataCollector.kt
package stasis.test.client_android.lib.mocks

import stasis.client_android.lib.collection.RecoveryMetadataCollector
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.TargetEntity
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger

class MockRecoveryMetadataCollector(val metadata: Map<Path, EntityMetadata>) : RecoveryMetadataCollector {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.FileCollectedWithExistingMetadata to AtomicInteger(0)
    )

    override suspend fun collect(
        entity: Path,
        destination: TargetEntity.Destination,
        existingMetadata: EntityMetadata
    ): TargetEntity {
        stats[Statistic.FileCollectedWithExistingMetadata]?.getAndIncrement()

        return when (val fileMetadata = metadata[entity]) {
            null -> throw IllegalArgumentException("No metadata found for file [$entity]")
            else -> TargetEntity(
                path = entity,
                destination = destination,
                existingMetadata = existingMetadata,
                currentMetadata = fileMetadata
            )
        }
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object FileCollectedWithExistingMetadata : Statistic()
    }
}
